server/migrations/1613615463607_reassign-participant-regions.js

```diff
@@ -0,0 +1,71 @@
+/* eslint-disable no-restricted-syntax, no-await-in-loop, max-len */
+const { readFileSync } = require('fs');
+const { join } = require('path');
+const readXlsxFile = require('node-xlsx').default;
+const { dbClient } = require('../db/db.js');
+const { collections } = require('../db/schema.js');
+
+const columnNames = ['ClientID', 'EOI - FHA', 'EOI - IHA', 'EOI - NHA', 'EOI - VCHA', 'EOI - VIHA'];
+const columnValues = ['Fraser', 'Interior', 'Northern', 'Vancouver Coastal', 'Vancouver Island'];
+
+// Program area provided a spreadsheet with participants' preferred health regions as columns
+// These columns contain textual quasi-booleans to denote whether the respective region is preferred
+// Booleans look like "YES" for truthy and "NO" or "NULL" for falsy (all case-insensitive)
+const textualBoolean = (v) => {
+  if (typeof v !== 'string') throw Error(`Invalid boolean: ${v} (type: ${typeof v})`);
+  switch (v.trim().toUpperCase()) {
+    case 'YES':
+      return true;
+    case 'NULL':
+    case 'NO':
+      return false;
+    default:
+      throw Error(`Invalid boolean string: ${v}`); // Better safe than sorry
+  }
+};
+
+exports.up = async () => {
+  const file = readFileSync(join(__dirname, 'assets', 'reassign-participant-regions.xlsx'));
+  const sheet = readXlsxFile.parse(file, { raw: true })?.[0]?.data; // Only need the first sheet
+  const [headers, ...data] = sheet; // First line is headers
+
+  if (!headers.every((e, i) => columnNames[i] === e)) throw Error('Unexpected headers');
+
+  const changes = data.map((row) => ({
+    id: row?.[0],
+    preferredLocation: row
+      .slice(1) // Ignore Maximus ID column
+      .map((v, i) => (textualBoolean(v) ? columnValues[i] : null)) // Map from textual boolean to HA string value
+      .filter((v) => v != null) // Remove null values where column was falsy
+      .join(';'), // For some reason, we are using a semicolon-separated string instead of a JSON array here
+  }));
+
+  const currentParticipants = await dbClient.db[collections.PARTICIPANTS].findDoc(
+    { 'maximusId::int': changes.map((i) => i.id) }, // Find all relevant participants to build history
+  );
+
+  for (const change of changes) {
+    const currentParticipant = currentParticipants.find((i) => i.maximusId === change.id);
+    // We could throw here if currentParticipant is undefined
+    // Currently, this is allowed and will set participant history item from value to null
+    // if (!currentParticipant) throw Error(`Could not find existing participant with Maximus ID ${change.id}`);
```

Comment:
On the fence re: enabling this safety check. With it, the migration will fail on dev and test. However, including it will catch the situation where a participant to be updated is not found in the current state of the DB. I think this is superfluous as I'd imagine the transaction used later would fail if this were the case.

Thanks for being my rubber duck. Fine to remove this check and comment if the team so decides.

Reply:
I don't think we need it. If one participant among many is missing, the outcome should just be that the missing record isn't updated - I'm not sure that should be a failure case, the migration succeeds as long as it updates any records that it actually does find :)

Reply:
Okay looking at your comments above - IMO we should just **continue** if currentParticipant is undefined, so keep this check but change the behaviour if hit

Reply:
`continue` would mean that `change.history` would be undefined. That would result in the participant still being updated but their `history` prop being set to `undefined`.

Reply:
The way it's currently written, the participant will still be updated (same as with `continue`) but will have a history array that correctly represents the change we're making. The missing `currentParticipant` is handled here using optional chaining. So if no current participant is found, `history` will look like `[{changes:[{field: 'preferredLocation', to: 'Fraser;Interior', from: null}], timestamp: '1234-56-78'}]`

Reply:
Discussed offline - we're good to remove this check 👍 the failing vs. ignoring missed records is handled on line 65, with the clause on line 66 taking care of it